File: src/blockset.ts
interface Block {
    // A 4x4 grid of tile IDs.
    tiles: number[][];
  }
  
  interface Blockset {
    blocks: Block[];
  }
  
  /**
   * Loads a blockset file (.bst) in 16-byte chunks.
   * Each block is exactly 16 bytes (read 4 rows of 4 numbers).
   *
   * NOTICE: Instead of copying the bytes with slice(),
   * we simply read them directly (and unroll the 4×4 grid).
   */
  export function loadBlockset(data: Uint8Array): Blockset {
    const numBlocks = (data.length / 16) | 0;
    const blocks = new Array<Block>(numBlocks);
    
    for (let i = 0; i < numBlocks; i++) {
      const base = i * 16;
      // Unroll the fixed 4×4 grid:
      blocks[i] = {
        tiles: [
          [data[base + 0], data[base + 1], data[base + 2], data[base + 3]],
          [data[base + 4], data[base + 5], data[base + 6], data[base + 7]],
          [data[base + 8], data[base + 9], data[base + 10], data[base + 11]],
          [data[base + 12], data[base + 13], data[base + 14], data[base + 15]]
        ]
      };
    }
    return { blocks };
  }
  
  /**
   * Loads a .blk file.
   * We assume that each byte in the Uint8Array represents a block index.
   * (This function's Array.from is typically fast enough, but if you want
   * even more speed you might work directly with the Uint8Array.)
   */
  export function loadBlk(data: Uint8Array): number[] {
    return Array.from(data);
  }
  
  /**
   * Assembles the full tile map from .blk and .bst data.
   * This version handles cases where the .blk file is smaller than the declared dimensions
   * by filling missing tiles with 0.
   */
  export function assembleMap(
    blkData: number[],
    blockset: Blockset,
    blocksW: number,
    blocksH: number
  ): number[][] {
    const totalRows = blocksH * 4;
    const totalCols = blocksW * 4;
    // Preallocate final 2D array:
    const finalMap: number[][] = new Array(totalRows);
    for (let r = 0; r < totalRows; r++) {
      finalMap[r] = new Array(totalCols).fill(0); // Initialize with 0s
    }
    
    const { blocks } = blockset;
    for (let by = 0; by < blocksH; by++) {
      for (let bx = 0; bx < blocksW; bx++) {
        const blkIndex = by * blocksW + bx;
        // Skip if we've run out of block data
        if (blkIndex >= blkData.length) {
          continue;
        }
        const blockIndex = blkData[blkIndex];
        if (blockIndex >= blocks.length) {
          // Out-of-range block index; you might handle this as needed.
          console.warn(`Warning: Block index ${blockIndex} is out of range; skipping.`);
          continue;
        }
        const block = blocks[blockIndex];
        // Compute where this block goes:
        const baseY = by * 4;
        const baseX = bx * 4;
  
        // With a 4×4 fixed grid we can simply "unroll" the inner loops:
        finalMap[baseY + 0][baseX + 0] = block.tiles[0][0];
        finalMap[baseY + 0][baseX + 1] = block.tiles[0][1];
        finalMap[baseY + 0][baseX + 2] = block.tiles[0][2];
        finalMap[baseY + 0][baseX + 3] = block.tiles[0][3];
  
        finalMap[baseY + 1][baseX + 0] = block.tiles[1][0];
        finalMap[baseY + 1][baseX + 1] = block.tiles[1][1];
        finalMap[baseY + 1][baseX + 2] = block.tiles[1][2];
        finalMap[baseY + 1][baseX + 3] = block.tiles[1][3];
  
        finalMap[baseY + 2][baseX + 0] = block.tiles[2][0];
        finalMap[baseY + 2][baseX + 1] = block.tiles[2][1];
        finalMap[baseY + 2][baseX + 2] = block.tiles[2][2];
        finalMap[baseY + 2][baseX + 3] = block.tiles[2][3];
  
        finalMap[baseY + 3][baseX + 0] = block.tiles[3][0];
        finalMap[baseY + 3][baseX + 1] = block.tiles[3][1];
        finalMap[baseY + 3][baseX + 2] = block.tiles[3][2];
        finalMap[baseY + 3][baseX + 3] = block.tiles[3][3];
      }
    }
    
    return finalMap;
  }
  
  /**
   * Prints the tile map to the console, formatting each tile as hex.
   */
  export function printMap(tilemap: number[][]): void {
    for (const row of tilemap) {
      console.log(row.map(tile => tile.toString(16).padStart(2, '0')).join(" "));
    }
  }
  
  /**
   * Combines loading and assembling the map.
   */
  export function loadAndAssembleMap(
    blkFileData: Uint8Array,
    blocksetFileData: Uint8Array,
    blocksW: number,
    blocksH: number
  ): number[][] {
    const blockset = loadBlockset(blocksetFileData);
    const blkData = loadBlk(blkFileData);
    const expected = blocksW * blocksH;
    if (blkData.length !== expected) {
      console.warn(`Warning: .blk file contains ${blkData.length} bytes but expected ${expected} (blocksW x blocksH)`);
    }
    return assembleMap(blkData, blockset, blocksW, blocksH);
  }
  
  /**
   * Gets a visual representation of the blockset data.
   */
  export function getBlocksetVisual(blocksetFileData: Uint8Array): number[][][] {
    const blockset = loadBlockset(blocksetFileData);
    return blockset.blocks.map(block => block.tiles);
  }

File: python_test/blockset.py
#!/usr/bin/env python3
"""
This script reads a Pokémon map’s block file (.blk) and its associated blockset (.bst)
and outputs a 2D array of base tile ids. (We ignore the actual graphic lookup here –
each number represents an 8×8 tile from the tileset.)

How it works:
  1. The blockset file contains a series of 16-byte blocks. Each block is a 4x4 grid
     of tile indices (each index refers to a tile from the 2bpp tileset image).
  2. The .blk file is a grid of bytes where each byte is an index into the blockset.
  3. Since each block is 4×4 base tiles and the map’s .blk file is given in “blocks” (for
     example, Pallet Town is 10 blocks wide by 9 blocks tall), the full tilemap will be
     (10×4) x (9×4) = 40×36 base-tile entries.
"""

import argparse

def load_blockset(filename):
    """
    Load a blockset file.

    Each block in the blockset is 16 bytes long and represents a 4x4 grid of base tile indices.
    Returns a list of blocks; each block is a list of 4 rows (lists of 4 integers each).
    """
    blocks = []
    with open(filename, "rb") as f:
        data = f.read()
    # Process data in 16-byte chunks:
    for i in range(0, len(data), 16):
        block_data = data[i:i+16]
        # If we get an incomplete block at the end, ignore it.
        if len(block_data) < 16:
            break
        block = []
        for row in range(4):
            row_tiles = []
            for col in range(4):
                index = row * 4 + col
                tile_id = block_data[index]
                row_tiles.append(tile_id)
            block.append(row_tiles)
        blocks.append(block)
    return blocks

def load_blk(filename):
    """
    Load a .blk file, which is simply a sequence of bytes.
    Each byte is an index into the blockset.
    Returns a list of integers.
    """
    with open(filename, "rb") as f:
        data = f.read()
    return list(data)

def assemble_map(blk_data, blocks, blocks_w, blocks_h):
    """
    Assemble the full tile map.

    The .blk file is a grid of block indices (dimensions: blocks_w × blocks_h).
    Each block from the blockset is 4×4 base tiles.
    The final tilemap will therefore have dimensions:
         width  = blocks_w * 4,
         height = blocks_h * 4.
    """
    # Initialize an empty 2D array (list of lists) for the final tile map.
    final_map = [[0 for _ in range(blocks_w * 4)] for _ in range(blocks_h * 4)]
    # Process each block reference in the .blk file:
    for by in range(blocks_h):
        for bx in range(blocks_w):
            # Calculate the index in the blk data
            blk_index = by * blocks_w + bx
            block_index = blk_data[blk_index]
            if block_index >= len(blocks):
                print(f"Warning: Block index {block_index} is out of range; skipping.")
                continue
            block = blocks[block_index]  # This is a 4x4 grid of tile ids.
            # Determine where this block goes in the final map:
            base_y = by * 4
            base_x = bx * 4
            for row in range(4):
                for col in range(4):
                    final_map[base_y + row][base_x + col] = block[row][col]
    return final_map

def print_map(tilemap):
    """
    Print the tilemap to the console. Each tile id is printed in hexadecimal.
    """
    for row in tilemap:
        print(" ".join(f"{tile:02X}" for tile in row))

def main():
    parser = argparse.ArgumentParser(
        description="Read a Pokémon .blk file and its blockset, then output a parent tilemap (2D array of tile ids)."
    )
    parser.add_argument("blk_file", help="Path to the .blk file (e.g. PalletTown.blk)")
    parser.add_argument("blockset_file", help="Path to the blockset file (e.g. overworld.bst)")
    parser.add_argument(
        "--blocks-width",
        type=int,
        default=10,
        help="Map width in blocks (default: 10 for Pallet Town)"
    )
    parser.add_argument(
        "--blocks-height",
        type=int,
        default=9,
        help="Map height in blocks (default: 9 for Pallet Town)"
    )
    args = parser.parse_args()

    blocks = load_blockset(args.blockset_file)
    blk_data = load_blk(args.blk_file)
    expected = args.blocks_width * args.blocks_height
    if len(blk_data) != expected:
        print(f"Warning: .blk file contains {len(blk_data)} bytes but expected {expected} (blocks_w x blocks_h)")

    tilemap = assemble_map(blk_data, blocks, args.blocks_width, args.blocks_height)
    print_map(tilemap)

if __name__ == '__main__':
    main()